pipeline/trusted-resources/pkg/webhook/validation_admit.go

```diff
@@ -0,0 +1,294 @@
+	cosignsignature "github.com/sigstore/cosign/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature"
+	sigstoresignature "github.com/sigstore/sigstore/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature/kms"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	admissionv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/apis"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/system"
+	"knative.dev/pkg/webhook"
+	"knative.dev/pkg/webhook/json"
+	"knative.dev/pkg/webhook/resourcesemantics"
+)
+
+var errMissingNewObject = errors.New("the new object may not be nil")
+
+const (
+	SecretPath      = "/etc/signing-secrets"
+	signatureSource = "tekton.dev/signature"
+	publicKey       = "tekton.dev/key"
+	kmsKey          = "tekton.dev/kms"
+)
+
+var _ webhook.AdmissionController = (*reconciler)(nil)
+
+// Admit implements AdmissionController
+func (ac *reconciler) Admit(ctx context.Context, request *admissionv1.AdmissionRequest) *admissionv1.AdmissionResponse {
+
+	kind := request.Kind
+	gvk := schema.GroupVersionKind{
+		Group:   kind.Group,
+		Version: kind.Version,
+		Kind:    kind.Kind,
+	}
+
+	ctx, resource, err := ac.decodeRequestAndPrepareContext(ctx, request, gvk)
+	if err != nil {
+		return webhook.MakeErrorStatus("decoding request failed: %v", err)
+	}
+
+	if err := ac.verifyResources(ctx, resource, gvk); err != nil {
+		return webhook.MakeErrorStatus("validation failed: %v", err)
+	}
+
+	return &admissionv1.AdmissionResponse{Allowed: true}
+}
+
+func (ac *reconciler) verifyResources(ctx context.Context, resource resourcesemantics.GenericCRD, gvk schema.GroupVersionKind) error {
+	logger := logging.FromContext(ctx)
+	logger.Info("Verifying Resources")
+
+	switch gvk.Kind {
+	case pipeline.TaskRunControllerName:
+		tr := resource.(*v1beta1.TaskRun)
+		if tr.ObjectMeta.Annotations == nil {
+			return fmt.Errorf("Annotations shouldn't be empty.")
+		}
+
+		if tr.ObjectMeta.Annotations[signatureSource] == "" {
+			return fmt.Errorf("%s shouldn't be empty.", signatureSource)
+		}
+
+		signatureEncoding := tr.ObjectMeta.Annotations[signatureSource]
+		signature, err := base64.StdEncoding.DecodeString(signatureEncoding)
+		if err != nil {
+			return err
+		}
+
+		var verifier sigstoresignature.Verifier
+		if tr.ObjectMeta.Annotations[kmsKey] != "" {
+			// Fetch key from kms
+			verifier, err = kms.Get(ctx, tr.ObjectMeta.Annotations[kmsKey], crypto.SHA256)
+			if err != nil {
+				return err
+			}
+		} else if tr.ObjectMeta.Annotations[publicKey] != "" {
+			// Fetch public key from annotation
+			cosignPublicKeypath := filepath.Join(tr.ObjectMeta.Annotations[publicKey], "cosign.pub")
+			verifier, err = cosignsignature.LoadPublicKey(ctx, cosignPublicKeypath)
+			if err != nil {
+				return err
+			}
```

Comment:
A local path on a remote resource is problematic, since this depends on the remote state to match what was done locally at signing.

This might make more sense if this was a reference to a secret?